ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -191,16 +249,87 @@ func (r *RayServiceReconciler) updateState(ctx context.Context, rayServiceInstan
-	// Update ray cluster
-	var rayClusterInstanceName string
+func (r *RayServiceReconciler) reconcileRayCluster(ctx context.Context, rayServiceInstance *rayv1alpha1.RayService) (*rayv1alpha1.RayCluster, *rayv1alpha1.RayCluster, error) {
+	rayClusterList := rayv1alpha1.RayClusterList{}
+	filterLabels := client.MatchingLabels{common.RayServiceLabelKey: rayServiceInstance.Name}
+	var err error
+	if err = r.List(ctx, &rayClusterList, client.InNamespace(rayServiceInstance.Namespace), filterLabels); err != nil {
+		return nil, nil, err
+	}
+
+	// Clean up RayCluster instances.
+	for _, rayClusterInstance := range rayClusterList.Items {
+		if rayClusterInstance.Name != rayServiceInstance.Status.RayClusterName && rayClusterInstance.Name != rayServiceInstance.Status.PreparingRayClusterName {
+			r.Log.Info("reconcileRayCluster", "delete ray cluster", rayClusterInstance)
+			if err := r.Delete(ctx, &rayClusterInstance, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil {
+				return nil, nil, err
+			}
+		}
+	}
+
+	// Get running cluster and preparing cluster instances.
+	runningRayCluster := &rayv1alpha1.RayCluster{}
 	if rayServiceInstance.Status.RayClusterName != "" {
-		rayClusterInstanceName = rayServiceInstance.Status.RayClusterName
+		if err = r.Get(ctx, client.ObjectKey{Name: rayServiceInstance.Status.RayClusterName, Namespace: rayServiceInstance.Namespace}, runningRayCluster); err != nil {
+			runningRayCluster = nil
+		}
+	} else {
+		runningRayCluster = nil
+	}
+
+	preparingRayCluster := &rayv1alpha1.RayCluster{}
+	if rayServiceInstance.Status.PreparingRayClusterName != "" {
+		if err = r.Get(ctx, client.ObjectKey{Name: rayServiceInstance.Status.PreparingRayClusterName, Namespace: rayServiceInstance.Namespace}, preparingRayCluster); err != nil {
+			preparingRayCluster = nil
+		}
 	} else {
-		rayClusterInstanceName = fmt.Sprintf("%s%s%s", rayServiceInstance.Name, rayClusterSuffix, rand.String(5))
+		preparingRayCluster = nil
 	}
 
-	r.Log.Info("getOrCreateRayClusterInstance", "rayClusterInstanceName", rayClusterInstanceName)
+	runningConfigKey := r.generateConfigKey(rayServiceInstance, rayServiceInstance.Status.RayClusterName)
+	preparingConfigKey := r.generateConfigKey(rayServiceInstance, rayServiceInstance.Status.PreparingRayClusterName)
+	// Clean up RayCluster serve deployment configs.
+	for key := range r.ServeDeploymentConfigs.Items() {
+		if key == runningConfigKey || key == preparingConfigKey {
+			continue
+		}
+		r.Log.Info("reconcileRayCluster removing config", "runningConfigKey", runningConfigKey, "preparingConfigKey", preparingConfigKey, "removekey", key)
+		r.ServeDeploymentConfigs.Remove(key)
+	}
+
+	shouldGeneratePreparingRayClusterName := false
+
+	if rayServiceInstance.Status.PreparingRayClusterName == "" {
+		if runningRayCluster == nil || !reflect.DeepEqual(runningRayCluster.Spec, rayServiceInstance.Spec.RayClusterSpec) {
+			shouldGeneratePreparingRayClusterName = true
```

Comment:
I'm finding the logic around `shouldGeneratePreparingRayClusterName` kind of confusing. What's going on in this part?

Reply:
It splits into 2 parts:
1. Generate a cluster name
2. Create the cluster cr